File: 3. Linked Lists/BBG. Flatten a Multi-level Linked List/Solution.py
class MultiLevelListNode:
    def __init__(self, val=None, next=None, child=None):
        self.val = val
        self.next = next
        self.child = child


def flatten_multi_level_list(head: MultiLevelListNode) -> MultiLevelListNode:
    if not head:
        return None

    tail = head
    while tail.next:
        tail = tail.next

    
    curr = head
    while curr:
        if curr.child:
            tail.next = curr.child
            curr.child = None

            while tail.next:
                tail = tail.next

        curr = curr.next
    
    return head

File: 3. Linked Lists/BBG. Flatten a Multi-level Linked List/Solution.cpp
struct MultiLevelListNode {
    int val;
    MultiLevelListNode *next;
    MultiLevelListNode *child;
    MultiLevelListNode() : val(0), next(nullptr), child(nullptr) {}
    MultiLevelListNode(int x) : val(x), next(nullptr), child(nullptr) {}
    MultiLevelListNode(int x, MultiLevelListNode *next, MultiLevelListNode *child)
        : val(x), next(next), child(child) {}
};


MultiLevelListNode* flatten_multi_level_list(MultiLevelListNode* head) {
    if (!head) return nullptr;

    MultiLevelListNode* tail = head;
    while (tail->next) {
        tail = tail->next;
    }

    MultiLevelListNode* curr = head;
    while (curr) {
        if (curr->child) {
            tail->next = curr->child;
            curr->child = nullptr;

            while (tail->next) {
                tail = tail->next;
            }
        }
        curr = curr->next;
    }

    return head;
}
